src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php

```diff
@@ -60,20 +61,30 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
      * passed which optimizes the object loading for one of the Doctrine
      * mapper implementations.
      *
-     * @param ChoiceListFactoryInterface $factory      The factory for creating
-     *                                                 the loaded choice list
      * @param ObjectManager              $manager      The object manager
      * @param string                     $class        The class name of the
      *                                                 loaded objects
      * @param IdReader                   $idReader     The reader for the object
      *                                                 IDs.
+     * @param ChoiceListFactoryInterface $factory      The factory for creating
+     *                                                 the loaded choice list
      * @param null|EntityLoaderInterface $objectLoader The objects loader
      */
-    public function __construct(ChoiceListFactoryInterface $factory, ObjectManager $manager, $class, IdReader $idReader = null, EntityLoaderInterface $objectLoader = null)
+    public function __construct($manager, $class, $idReader = null, $objectLoader = null, $factory = null)
     {
+        // BC to be removed in 4.0
+        if (null !== $this->factory && $objectLoader instanceof EntityLoaderInterface) {
```

Comment:
For clarity, I'd change the `instanceof` check to `$manager instanceof ChoiceListFactoryInterface`. I'd also add a comment that we need to add back the type hints in 4.0.